transports/onion/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+    pub fn builder() -> OnionBuilder<tor_rtcompat::tokio::TokioRustlsRuntime> {
+        let runtime = tor_rtcompat::tokio::TokioRustlsRuntime::current()
+            .expect("Couldn't get the current tokio rustls runtime");
+        TorClient::with_runtime(runtime)
+    }
+}
+
+#[cfg(all(feature = "native-tls", feature = "async-std"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "native-tls", feature = "async-std"))))]
+pub type OnionAsyncStdNativeTlsClient =
+    OnionClient<tor_rtcompat::async_std::AsyncStdNativeTlsRuntime>;
+#[cfg(all(feature = "rustls", feature = "async-std"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "rustls", feature = "async-std"))))]
+pub type OnionAsyncStdRustlsClient = OnionClient<tor_rtcompat::async_std::AsyncStdRustlsRuntime>;
+#[cfg(all(feature = "native-tls", feature = "tokio"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "native-tls", feature = "tokio"))))]
+pub type OnionTokioNativeTlsClient = OnionClient<tor_rtcompat::tokio::TokioNativeTlsRuntime>;
+#[cfg(all(feature = "rustls", feature = "tokio"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "rustls", feature = "tokio"))))]
+pub type OnionTokioRustlsClient = OnionClient<tor_rtcompat::tokio::TokioRustlsRuntime>;
+
+#[derive(Debug, Clone, Copy, Default)]
+pub struct AlwaysErrorListenerUpgrade;
+
+impl core::future::Future for AlwaysErrorListenerUpgrade {
+    type Output = Result<OnionStream, OnionError>;
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        panic!("onion services are not implented yet, since arti doesn't support it. (awaiting Arti 1.2.0)")
+    }
+}
+
+impl<R: Runtime> Transport for OnionClient<R> {
+    type Output = OnionStream;
+    type Error = OnionError;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+    type ListenerUpgrade = AlwaysErrorListenerUpgrade;
+
+    /// Always returns `TransportError::MultiaddrNotSupported`
+    fn listen_on(
+        &mut self,
+        addr: libp2p_core::Multiaddr,
+    ) -> Result<
+        libp2p_core::transport::ListenerId,
+        libp2p_core::transport::TransportError<Self::Error>,
+    > {
+        // although this address might be supported, this is returned in order to not provoke an
+        // error when trying to listen on this transport.
+        Err(TransportError::MultiaddrNotSupported(addr))
+    }
+
+    /// Always returns false
+    fn remove_listener(&mut self, _id: libp2p_core::transport::ListenerId) -> bool {
+        false
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let tor_address = if !self.safe_mode {
+            safe_extract_tor_address(&addr).or_else(|_| dangerous_extract_tor_address(&addr))
+        } else {
+            safe_extract_tor_address(&addr)
+        }
+        .map_err(OnionError::from)
+        .map_err(TransportError::Other)?;
```

Comment:
Here we should return a `TransportError::MultiaddrNotSupported` in case we can't dial the address.

Reply:
Resolved in latest commit.